src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Hosting/VSInternalServerCapabilitiesExtensions.cs

```diff
@@ -57,4 +61,33 @@ public static void EnableMapCodeProvider(this VSInternalServerCapabilities serve
     {
         serverCapabilities.MapCodeProvider = true;
     }
+
+    public static HashSet<string> HtmlAllowedAutoInsertTriggerCharacters { get; } = new(StringComparer.Ordinal) { "=", };
+    public static HashSet<string> CSharpAllowedAutoInsertTriggerCharacters { get; } = new(StringComparer.Ordinal) { "'", "/", "\n" };
```

Comment:
Seems odd to have these here. I would have thought putting them on `OnAutoInsertService` would make more sense. Callers could do their own concatenation with `triggerCharacters` (in cohosting it always concats, in the existing server it checks the `SingleServer` flag), and then just pass them in to the methods below.

Reply:
This seemed like the right place to me since this is the common code for the capabilities and those are part of capabilities data. I followed your suggestion and moved them to individual handlers but I'm not sure I like that better. We end up with less common code and more (what seems to be) duplicate code/logic. Curious to see if you like the new change better.

Reply:
The way I read it now, the only duplicated code is a single `Concat` call. To me it makes perfect sense that the logic would be in the individual endpoints, given they're not the same (one needs to check a feature flag, one doesn't). The characters themselves are still defined centrally, and more importantly, they're defined in a class that is specific to AutoInsert, and not in a class specific to a particular LSP protocol type library.